python/ray/autoscaler/_private/spark/node_provider.py

```diff
@@ -0,0 +1,210 @@
+        return node_id
+
+    def external_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def internal_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def set_node_tags(self, node_id, tags):
+        assert node_id in self._nodes
+        self._nodes[node_id]["tags"].update(tags)
+
+    def create_node(self, node_config: Dict[str, Any], tags: Dict[str, str], count: int) -> Optional[Dict[str, Any]]:
+        raise AssertionError("This method should not be called.")
+
+    def _gen_spark_job_group_id(self, node_id):
+        return f"ray-cluster-{self.ray_head_port}-{self.cluster_unique_id}-worker-node-{node_id}"
+
+    def create_node_with_resources_and_labels(
+        self, node_config, tags, count, resources, labels
+    ):
+        for _ in range(count):
+            with self.lock:
+                resources = resources.copy()
+                node_type = tags[TAG_RAY_USER_NODE_TYPE]
+                next_id = self.get_next_node_id()
+                resources["NODE_ID_AS_RESOURCE"] = next_id
+
+                """
+                ray_worker_node_cmd = [
+                    sys.executable,
+                    "-m",
+                    "ray.util.spark.start_ray_node",
+                    f"--temp-dir=/tmp/test_worker1",
+                    f"--num-cpus={resources.pop('CPU', 0)}",
+                    f"--num-gpus={resources.pop('GPU', 0)}",
+                    "--block",
+                    "--address=192.168.10.116:3344",
+                    f"--memory={512 * 1024 * 1024}",
+                    f"--object-store-memory={resources.pop('object_store_memory', 512 * 1024 * 1024)}",
+                    f"--min-worker-port={11000}",
+                    f"--max-worker-port={20000}",
+                    f"--resources={json.dumps(resources)}",
+                    # f"--dashboard-agent-listen-port={ray_worker_node_dashboard_agent_port}",
+                    # *_convert_ray_node_options(worker_node_options),
+                ]
+                
+                from ray.util.spark.cluster_init import setup_sigterm_on_parent_death, exec_cmd, RAY_ON_SPARK_COLLECT_LOG_TO_PATH
+                ray_node_proc, tail_output_deque = exec_cmd(
+                    ray_worker_node_cmd,
+                    synchronous=False,
+                    preexec_fn=setup_sigterm_on_parent_death,
+                    extra_env={
+                        RAY_ON_SPARK_COLLECT_LOG_TO_PATH: "",
+                        "RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER": "1",
+                    },
+                )
+                """
+                conf = self.provider_config
+
+                num_cpus_per_node = resources.pop('CPU')
+                num_gpus_per_node = resources.pop('GPU')
+                heap_memory_per_node = resources.pop('memory')
+                object_store_memory_per_node = resources.pop('object_store_memory')
+                response = requests.request(
+                    url=self.server_url + "/create_node",
```

Comment:
@jjyao 

To confirm the behavior, assuming creating a worker node taking about 5 mins, but here it just sending the creating node request and then returns immediately. Does this way work well with autoscaler framework ?

Reply:
I think it's fine.

Reply:
@jjyao 

But assuming the case: after sending worker node creation requests, 5 mins later, the node creation failed, then how does autoscaler handle the case ? Does it trigger the routine you mentioned here ? https://github.com/ray-project/ray/pull/38215#discussion_r1295417619